### ISSUE TITLE
Filter out non-functional beans before applying validation checks

### ISSUE THREAD

**Issue by elefeint** [spring-cloud/spring-cloud-function#417]
The fix in #409 addressed the issue with `Supplier`/`Function`/`Consumer` beans ending up in the function registry in applications that were not meant to be functional.

There is another scenario: applications that do intend to be functional, and provide a proper `Supplier`/`Function`/`Consumer` bean, but also happen to declare (or pull in through dependencies) one or more accidentally functional beans. These would be [identified as valid](https://github.com/spring-cloud/spring-cloud-function/blob/fa1b9c69c3deb9ca951cd2ea536fdd8812411ecf/spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/catalog/BeanFactoryAwareFunctionRegistry.java#L210-L215) `Supplier`/`Function`/`Consumer` at the first stage. If only a single accidentally functional bean existed, it would get caught in the [declared type validations](https://github.com/spring-cloud/spring-cloud-function/blob/fa1b9c69c3deb9ca951cd2ea536fdd8812411ecf/spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/catalog/BeanFactoryAwareFunctionRegistry.java#L241-L248). But if one **real** and one or more accidental functional beans exist, the logic never gets that far because it trips up a [validation for the number of functional beans](https://github.com/spring-cloud/spring-cloud-function/blob/fa1b9c69c3deb9ca951cd2ea536fdd8812411ecf/spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/catalog/BeanFactoryAwareFunctionRegistry.java#L224), which results in the function definition never being added to the function registry.

From a user application perspective, a warning message "Found more then one function beans in BeanFactory" is logged, and the binder is never activated. User applications would have to set `spring.cloud.function.definition` explicitly to get correct behavior.

One possible solution is to prevent "accidentally functional" beans from becoming candidates for autodiscovery. For example, this [validation block of code ](https://github.com/spring-cloud/spring-cloud-function/blob/fa1b9c69c3deb9ca951cd2ea536fdd8812411ecf/spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/catalog/BeanFactoryAwareFunctionRegistry.java#L241-L248) could become a filter to be applied [while aggregating functions](https://github.com/spring-cloud/spring-cloud-function/blob/fa1b9c69c3deb9ca951cd2ea536fdd8812411ecf/spring-cloud-function-context/src/main/java/org/springframework/cloud/function/context/catalog/BeanFactoryAwareFunctionRegistry.java#L221), before the check for the number of autodiscovered functions is done.

For additional context, see spring-cloud/spring-cloud-gcp#1997 and [this sample application](https://github.com/spring-cloud/spring-cloud-gcp/tree/master/spring-cloud-gcp-samples/spring-cloud-gcp-pubsub-stream-binder-functional-sample) that works because it sets the `spring.cloud.function.definition` bean.

**Comment by olegz** [comment]
@elefeint While this is a valid observation, in reality I do not force a production app where `spring.cloud.function.definition` is not specified. Also, there will always be a scenario where a perfectly valid function is not intended to be used in s-c-f scope, so no meter how many filters we may add a new edge case will pop up and so on.
The auto-discovery feature was never meant to be some form of magic. It was always _if you have one we got it otherwise . . ._

In other words i'll let this issue sit for a while to see if there are other opinions, but at the moment i don't foresee any action.

**Comment by elefeint** [comment]
It makes sense to wait and see if there are other projects affected. We'll proceed with spring-cloud/spring-cloud-gcp#1997 as a workaround for now.

**Comment by olegz** [comment]
@elefeint Given that variations or issue did indeed pop up quite a few times, most if not all possible cases have been addressed to avoid what happened with you in GCP and others. Also, it is very relevant to #418 

So, I am going to close this for now as there is nothing i can see us doing, but certainly if something pops up again we can reopen or raise a new issue.
